rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,277 @@
+foundation, we can set the stage to migrate current GraphQL API use cases
+towards technology more suited to the team maintaining it.
+
+### Make providers more synchronous
+
+Likely the smallest change we could make would be to allow a feedback mechanism
+for providers. This would allow us to implement the control server as a
+provider with fewer structural changes to Vector's main run loop. Concretely,
+this could look something like adding an optional oneshot response channel to
+the `SignalTo::ReloadFromConfigBuilder` variant used by providers.
+
+We don't take this route because the signal-based focus of reloading is already
+overloaded enough to make the process difficult to follow. It's not a good
+match for what we're trying to do and would like further obscure the logic,
+making bugs more difficult to track down. In fact, given the limited uptake of
+the provider system in general, it's possible that it should be reworked to fit
+the same pattern as the new control server. This could be done as future work.
+
+### More isolated control socket mode
+
+Another option would be to branch earlier in `Application::run`, make options
+related to the control server exclusive with normal options like `--config`,
+and have more of a separate mode of operation for this functionality. This
+would remove some concerns about potential strange interactions between
+multiple paths for reloading (e.g. watching a file while running a control
+server while running an HTTP provider) and potentially require less refactoring
+work than what's proposed in this RFC.
+
+This approach wasn't chosen because the divergence for "normal" Vector
+operation would itself be a path for potentially confusing differences in
+behavior. It also would not move us closer to an improved overall design for
+this portion of Vector, and would create another path that we would later
+likely want to unify.
+
+### gRPC instead of JSON
+
+Instead of a REST-ish JSON API, we could use gRPC. We already use the
+technology in other areas of Vector (native source and sink), so it wouldn't
+have the same downside as GraphQL, while providing some of the benefits of an
+IDL.
+
+This approach was not chosen for a couple of reasons:
+- Protobuf adds complexity (i.e. code generation) to any client implementation.
+- The ability to interact with the API from simple tools like cURL is lost.
+- The only current endpoint accepts a Vector config, for which we already have
+  a JSON Schema and we would not want to translate to Protobuf.
+- Many of the benefits of gRPC are around high performance, load balancing,
+  health checking, etc, none of which is relevant to a small local control API.
+
+## Outstanding Questions
+
+- What is the expected flow for setting global options at startup (especially
+  those that can't be changed via reload)? Use normal file loading with only
+  global options set?
+- Should the API accept entire configs, or should we attempt to limit to just
```

Comment:
Somewhat outside the scope of this particular question, but I'm starting to wonder what exactly the UX for reloading looks like at the OPW level? Are we going to be replicate the `SIGHUP`/file watching behavior of Vector? 

Or am I thinking about this wrong, and this reloading API is only intended for remote config use cases?

Reply:
I'd be thinking that OPW wouldn't have any file-based reloading (any FS and/or env/flag changes would be a side effect of how the service is deployed -- k8s, systemd, etc). The only "reloading" would be RC which would remain in process.

I'd like to get @lukesteensen's thoughts on this to confirm though.

Reply:
The assumption at this point is that users are going to doing either one or the other. If they're using local config files and sending SIGHUP, that'll continue to work as it does today (for now, see below). We don't currently let them set the file watching flag on Vector via OPW, so that's out of the picture for now. Then if they're using remote config we assume for now that they aren't doing things like sending SIGHUP, which would cause odd behavior in the current (proposed) state.

This is a good question though, because it points out a few things that we are going to need to decide at a product level. What do we want to behavior to be here, and how do we expect users to want to operate it? The point of the API as the RFC stands is to get feedback from reloads, which is most relevant for remote config. What should SIGHUP do in that scenario? It'd be relatively simple to "deactivate" reloading from disk on SIGHUP after configs have been loaded via the API, but that's not a very principled approach.

Ultimately, as mentioned in the future work section, I think we're going to need to make some decisions around separating "types" of config and how they're presented to solve the bootstrapping problem, then use that distinction to more clearly separate different modes of operation (i.e. API-only vs file-only). Once the modes are established, it should be easier to answer questions like what SIGHUP should do in specific cases. This RFC opts to make forwards progress without blocking on those decisions, but we should still try to answer them as soon as possible.